crates/ruff_python_parser/tests/snapshots/invalid_syntax@star_index_py310.py.snap

```
@@ -0,0 +1,420 @@
+                                                            range: 262..263,
+                                                            value: Int(
+                                                                5,
+                                                            ),
+                                                        },
+                                                    ),
+                                                ),
+                                                step: None,
+                                            },
+                                        ),
+                                        Starred(
+                                            ExprStarred {
+                                                range: 265..270,
+                                                value: Name(
+                                                    ExprName {
+                                                        range: 266..270,
+                                                        id: Name("idxs"),
+                                                        ctx: Load,
+                                                    },
+                                                ),
+                                                ctx: Load,
+                                            },
+                                        ),
+                                    ],
+                                    ctx: Load,
+                                    parenthesized: false,
+                                },
+                            ),
+                            ctx: Load,
+                        },
+                    ),
+                },
+            ),
+        ],
+    },
+)
+```
+## Unsupported Syntax Errors
+
+  |
+1 | # parse_options: {"target-version": "3.10"}
+2 | lst[*index]  # simple index
+  |     ^^^^^^ Syntax Error: Cannot use star expression in index on Python 3.10 (syntax was added in Python 3.11)
```

I'm not sure this is correct. I haven't checked if it is a parse or semantic error but Python 3.12 gives me:

```
>>> [][*list]
<python-input-1>:1: SyntaxWarning: list indices must be integers or slices, not tuple; perhaps you missed a comma?
  [][*list]
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    [][*list]
    ~~^^^^^^^
TypeError: list indices must be integers or slices, not tuple
```

Which is also what I read from the grammar change: It only allows star expressions for slices except the first

---

Yeah, I think the check needs to be moved up in the `if` branch which is a pure tuple case.

This is a `TypeError` and is not relevant to the parser:
```console
$ bat _.py                         
         File: _.py
     1 ~ [][*data]

$ uv run --python=3.12 --no-project python -m ast _.py
Module(
   body=[
      Expr(
         value=Subscript(
            value=List(elts=[], ctx=Load()),
            slice=Tuple(
               elts=[
                  Starred(
                     value=Name(id='data', ctx=Load()),
                     ctx=Load())],
               ctx=Load()),
            ctx=Load()))],
   type_ignores=[])
```

---

I may have made this test case a bit misleading by naming the container `lst`, but as @dhruvmanila points out with the valid AST, this is *syntactically* valid, and it's *not* a type error for a container like a numpy array or another container that can be validly indexed by a tuple:

```pycon
>>> import numpy as np
>>> a = np.array([[1, 2, 3], [4, 5, 6]])
>>> t = [1, 2]
>>> a[*t]
np.int64(6)
```

It's not a parser or compiler error, it's only a runtime `TypeError` if the container doesn't support tuple indexing.

I believe the grammar in the [reference](https://docs.python.org/3/reference/expressions.html#slicings) also supports this, albeit with different terminology from the PEP:

```
slicing      ::= primary "[" slice_list "]"
slice_list   ::= slice_item ("," slice_item)* [","]
slice_item   ::= expression | proper_slice
proper_slice ::= [lower_bound] ":" [upper_bound] [ ":" [stride] ]
lower_bound  ::= expression
upper_bound  ::= expression
stride       ::= expression
```

`slice_item` is either an `expression` *or* a `proper_slice` with multiple parts.

---

Sorry, I think I mis-understood the change here. I think what Brent has here is correct. The example and I and Micha have pointed out is a runtime error and should throw a syntax error for Python version older than 3.11.